### DISCUSSION TITLE
[Suggestion] Wishing Well Revamp

### DISCUSSION THREAD

**Discussion by eliasSFL** [sunflower-land/sunflower-land#3512] [Ideas]
# Current Flow for Wishing Well

1. User interacts with WW Contract, 
- WW Tokens are minted into the user's wallet 1:1 to the amount of Quickswap MATIC-SFL Liquidity Tokens the user has. 
- MATIC-SFL Tokens stay in user's wallet
2. Every 3 days user can claim SFL Rewards into the user's farm. 
- The reward given is 3 days worth of tokens, regardless of how many days from the last claim date. 
- If Liquidity is removed Partially, WW tokens equivalent to the amount removed is burnt
3. The user removes MATIC-SFL Liquidity entirely. 
- WW Tokens are not burnt, but the user won't be able to earn further rewards.

# Issues with the current WW System
1. User's WW Tokens aren't burnt when the user removes MATIC-SFL Liquidity entirely from the user's wallet
- This creates a situation where there are unused WW tokens that are contributing to the calculation of rewards the user gets. This results in players getting lesser SFL Rewards from WW than they are supposed to get.

For reference: 
1,180,849.096892 WW Tokens minted in the Blockchain
350,588.32126 MATIC-SFL Liquidity Tokens minted
This results in 830260.775632 wasted WW tokens sitting in user's wallet not collecting rewards but it's contributing to the total supply of WW, which the rewards are calculated.
There are currently 50,000 SFL in the WW pool
If a user deposits 100 USD ($50 MATIC and $50 SFL) they will get 158.743 MATIC-SFL Liquidity Tokens, when interacting with WW contract, they will mint 158.743 WW tokens
Current the user will get the following amount of tokens every 3 days:
(158.743/1,180,849.096892)*50,000 SFL = 6.72156 SFL

If the total is based on the actual amount of liquidity tokens it would look more like this:
(158.743/350,588.32126)*50,000 SFL = 22.639515 SFL
This is 237% more SFL than what we are earning right now.

2. Rewards earned is based on the share of WW tokens rather than share of WW tokens over time
As I've calculated above, the rewards you get every 3 days is based on the share of WW tokens you have, time isn't considered when calculating rewards. As a result, if a user claims 5 days after they last claimed, they still get 3 days worth of SFL instead of 5 days worth of rewards
((158.743/350,588.32126)*(50,000/3))x5 = 37.732 SFL if user claims 5 days later

So I suggest the formula to be calculated as such
SFL Rewards = ([Amount of WW]/[Total amount of WW])x([Amount of SFL in contract]/3)x[Number of days since last claim]

# Suggestion to improve WW

1. User deposits liquidity
- MATIC-SFL LP sent to WW Contract
- WW Tokens equivalent to the amount of MATIC-SFL LP minted
2. Rewards
- Accured over time
- Can be claimed every 3 days, but calculated based on how long since last claim
3. User removes liquidity
- WW Tokens gets burnt
- User gets rewards from the time of last claim
- MATIC-SFL LP tokens returns to user's wallet

Do let me know if you guys have any opinions or suggestions to improve Wishing Well

**Comment by rodolfoddias** [comment]
Suggestions:

- Make it claimable through the farm itself, something like the daily chest! (maybe we could build or own WW inside our farm)
- Let players provide liquidity through the game, not outside of it. 

I would love to provide liquidity, but I give up due the withdrawing taxes most of the times.

**Comment by adamhannigan** [comment]
Thanks for the write up @eliasSFL This is something we are looking to implement. It will take roughly 3-4 days of dedicated team work - we just need to find the time for it 😅 

Along with your suggestions and @rodolfoddias we will try incorporate to a more reliable and automated system. Ideally making it easier for LPers to claim rewards (such as off chain claiming and chests on main farm)

**Reply by eliasSFL** [reply]
Hey @adamhannigan thanks for the reply! Some of the community have discussed this with more suggestions, when u have some time, u can have a look at it: https://discord.com/channels/880987707214544966/1217387658943135744
